### PR TITLE
Feature/adding support for artist

### LINKED ISSUE #51: Tried to download a song, the song is not on youtube and it downloaded the first result, which is a 4.5 hour compilation
Tried to download https://open.spotify.com/track/5FAm2NCjU38DdARBhdwEzT
What was downloaded: https://www.youtube.com/watch?v=fbDwaUej4aE
I tried searching for the song on youtube, but I couldn't find it, so I guess that's why, but still, there should be a limit/confirmation for song length, maybe error if it's not found, since it was stuck at 0MB out of 0MB until it finished, and the result file is over 500MB.

### LINKED ISSUE #75: Download Problem
So during downloading a playlist, songs seem to disappear from the folder. For example I have 200 songs in one playlist, although the command line says everything is full downloaded, my folder info only shows 92 items. Why is this so, is this a bug? Hope you can address the issue.

This is the playlist that doesn't work (https://open.spotify.com/playlist/6J08oqfmAdKRrRlkIMR5LK?si=yjU4g-uUTcWKypvQ1iymXQ), Its quite long though.

### LINKED ISSUE #80: [feature] download a song using YT link directly from CLI
This sounds like its outside the scope of this package but hear me out.
We already have all the necessary stuff to be able to do this and i found a song not on spotify but available on YT
I could download and use yt-dl separately but having this option here feels convenient

### LINKED ISSUE #82: Second pull of a spotify playlist returns undefined/ error code
Environment

docker (debian)

master up-to-date

Install Source:
npm/ dockerfile in source

Expected Behavior

i can re run spotifiy downloader for the same url over and over and get a success exit code in command line
Actual Behavior

actual behavior 
it returns "undefined" and throws a SIGTERM or ERR
Steps to Reproduce

```
#!/bin/bash
trap 'echo "oh no!"' ERR SIGTERM

declare -A sources=(
        ["test"]="https://open.spotify.com/playlist/3ZvfFPjhl8RoL2HNNV0uqY?si=g2zMkE2NRzu_QC8vabRCDA"
)
    spotifyPath="/tmp"
    for sourceKey in "${!sources[@]}";
    do
        source=${sources[$sourceKey]}
        
        echo "Downloading ${sourceKey}"
        docker run -u $(id -u):$(id -g) -v ${spotifyPath}:/download:rw --rm phyzical/spotify-dl "$source"
    done
    
    echo "Finished Spotify Download!!"
```
above is the bash script, first run will download and return a valid exit code
if you re run for the same playlist url i get the following

```
Sun Apr 11 18:40:01 AWST 2021 Downloading test2
- Searching…
ℹ Total Songs: 3
ℹ Saving Playlist: /download/Silver-Test - Jayden
ℹ 1. Song: Blinding Lights - The Weeknd
- Downloading...
✔ Download completed.
✔ Metadata Merged!
ℹ 2. Song: Hello - Adele
- Downloading...
✔ Download completed.
✔ Metadata Merged!
ℹ 3. Song: Never Gonna Give You Up - Rick Astley
- Downloading...
✔ Download completed.
✔ Metadata Merged!
Sun Apr 11 18:40:29 AWST 2021
Sun Apr 11 18:40:29 AWST 2021 Finished. Saved 3 Songs at /download/Silver-Test - Jayden/Never Gonna Give You Up - Rick Astley.mp3.
Sun Apr 11 18:40:30 AWST 2021 Finished Spotify Download!!
Script Finished Apr 11, 2021 18:40.31

Full logs for this script are available at /tmp/user.scripts/tmpScripts/spotify downloader/log.txt

Script Starting Apr 11, 2021 18:40.36

Full logs for this script are available at /tmp/user.scripts/tmpScripts/spotify downloader/log.txt

Sun Apr 11 18:40:36 AWST 2021 Downloading test2
- Searching…
ℹ Total Songs: 3
ℹ Saving Playlist: /download/Silver-Test - Jayden
ℹ Fetching cache to resume Download

undefined
Sun Apr 11 18:40:39 AWST 2021 Finished Spotify Download!!
Script Finished Apr 11, 2021 18:40.39
```

the reason i hoped it would return a similar exit code is i trap on SIGTERM or ERR to send notifications of an issue when running these

also note the trap isnt hit the first time only on subsequent attempts of the playlist

thanks in advance!

### LINKED ISSUE #83: add support for get all albums for an artist
i have been playing with this along with a fix for my last issue aswell as some drying up of the code 👍

### LINKED ISSUE #86: WARNING: ytdl-core is out of date! Update with "npm install ytdl-core@latest"
Where should I call this? I tried many folders: `c:\program files\nodejs`,  `%AppData%\npm\node_modules`, as global command

It throws warnings and errors and then still show message that ytdl is out of date. How do I fix this?

### PR DESCRIPTION
**Pull request by phyzical**
pile of changes sorry if its too much,

fixes #83 
fixes #82 
fixes #86 
fixes #80 
fixes #51 
fixes #75 

i have confirmed that each of the following still work after the changes aswell as testing the "page" sizes by manually setting the limits so i think i have tested fairly well
["artist"]="https://open.spotify.com/artist/0DGqabk3gsZWS8zEdgt1m5?si=ECdth_rMSnafUkGLskc08g"
["album"]="https://open.spotify.com/album/3XtEGVx9uh7J46nBzEc1VS?si=hUiwIxK0SZeCtWlEMUbmPA"
["playlist"]="https://open.spotify.com/playlist/46HnjD7FcxjwcTnbtoDVxI?si=je1jJpTvRCOl9jvo_tB8TA"  ["song"]="https://open.spotify.com/track/2wBjgUXl43nSK7k3jVrSac?si=O6GOeMIkRyy9q-bBBYGZJQ"
["youtube"]="https://youtube.com/watch?v=372obTG80y0"
  ["playlist-2-#86"]="https://open.spotify.com/playlist/1rPNGZ0mUBNrzIY9mgKhgW?si=7MYsq9ZbRD6x3Uifv9d-cg"

rough summary of changes include
* linting changes
* support for eslint and editor config (other forms of linting)
* added support to download all album's songs for an artist
* DRY'd up a bunch of common code between the multiple methods
* refactored some function/class/variable names for increased readability
* improved caching system to save ids of spotify songs instead of numbers
* added support for direct youtube links
* added a max length of around 1 hour per song

### CODE REVIEW

**Comment by phyzical** [comment]
Also fixes #86 an #80

**Comment by phyzical** [comment]
also fixes #51

**Comment by SwapnilSoni1999** [comment]
Thankyou so much @phyzical loved this PR. Will you give me some time to merge and test the code. I also saw you mentioned your PR on other issues as well. :) I'll test and inform you back soon 
Thankyou once again.

**Comment by phyzical** [comment]
happy to help :) sorry about the feature creep though haha.. just got a bit carried away 😄 

also don't be afraid to question anything ive done 😉 

basically none of the other options (repo wise) out there seemed to have an easy docker setup + artist support so this one seemed easy enough to get going.

two things that i couldn't get down correctly though was
*  the downloading % text so atm it is separate info calls (means a bit more console log bloat)
* making it completely sync rather than async.. though i kind of prefer it as itll download things a little bit faster (just means the log is kind of confusing at times 😆 )
but both of the above would be solved by better logging management (unless i cause that regression with my changes  haha)

**Review by SwapnilSoni1999 — APPROVED** [review]

**Review by SwapnilSoni1999 — CHANGES_REQUESTED** [review]

**Comment by phyzical** [comment]
i actually think i just found the cause of #75 

when the user requests a link containing ?si= the web node api makes the url like this `https://api.spotify.com/v1/playlists/3oTMrnB9hH7Zw2Px1FWyYi?si=_VccSY9dSIOPCQFlmO6Yqg/tracks?limit=50&offset=400` and well its now obvious why the offset and limit is lost 👍 so its always defaulting to the api limit of 100 and only ever downloads 100 songs doh

ill post a fix for the above then address the comments

File: lib/downloader.js [hunk 1]
```diff
@@ -2,18 +2,29 @@
...
   const download = ytdl(youtubeLink, options);
+  spinner.start('Downloading...');
   download.on('progress', (chunk, downloaded, total) => {
-    spinner.text = `Downloading ${(downloaded / 1024 / 1024).toFixed(2)}MB of ${(total / 1024 / 1024).toFixed(2)}MB\n`;
+    const toBeDownloadedMb = (downloaded / 1024 / 1024).toFixed(2);
+    const downloadedMb = (total / 1024 / 1024).toFixed(2);
+    if (toBeDownloadedMb % 1 == 0) {
+      // spinner.text doesnt seem to work for me?
+      spinner.info(
+        `Downloaded ${toBeDownloadedMb}/${downloadedMb} MB - ${youtubeLink}`,
```
[the remark this answers is not in the document]
Comment (phyzical) [reply lib/downloader.js:25]: when you say the downloaded mb gets printed wrong do you mean that it never shows the last print i.e 2.35/2.35 MB

[the remark this answers is not in the document]
Comment (phyzical) [reply lib/downloader.js:25]: yeah ill remove the youtube link 👍

File: cli.js [hunk 2]
```diff
@@ -77,116 +77,139 @@ if (!input[0]) {
...
+          dir,
+          filter.validateOutputSync(
+            `${songInfo.name} - ${songInfo.artists[0]}.mp3`,
+          ),
+        );
+        spinner.info(`DIR: ${listData.name}`);
+        await download(ytLink, output, spinner, async () => {
+          await cache.write(path.join(dir, '.spdlcache'), trackId);
+          await mergeMetadata(output, songInfo, spinner, async () => {
+            listData.tracks = listData.tracks.map(track => {
```
[the remark this answers is not in the document]
Comment (phyzical) [reply cli.js:118]: sorry why would we want to ever remove the cache file?

**Comment by phyzical** [comment]
i found out how to promisify the ffmpeg and chuck a wait, so now its fully synchronous, this way it doesn't start to snow ball with multiple downloads for i.e artist downloaded. (an issue i was facing sometimes)

It will also make it easier to speed things up when we decide to async certain parts i.e just the metadata so the next dl can start while it writes ect.

**Comment by SwapnilSoni1999** [comment]
Amazing @phyzical <3 I'll check this out asap. Thankyou once again, You're doing great

**Comment by phyzical** [comment]
about to push one more fix.. haha now that things like "Artist" download take over an hour were hitting the access token expiry. im just testing an artist now with code to use the refresh token, should be pushed within the next 30 👍

**Comment by battlezone** [comment]
I'd like to test it but I cannot get it to run. I first installed Spotify-dl via NPM and encountered issue #75 and @phyzical replied to the issue to check out this PR. When trying to run I get the following error:
```
node:internal/modules/cjs/loader:944
  throw err;
  ^

Error: Cannot find module 'bl'
Require stack:
- /mnt/e/User/Music/Spotify/spotify-dl/node_modules/ora/index.js
- /mnt/e/User/Music/Spotify/spotify-dl/cli.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:941:15)
    at Function.Module._load (node:internal/modules/cjs/loader:774:27)
    at Module.require (node:internal/modules/cjs/loader:1013:19)
    at require (node:internal/modules/cjs/helpers:93:18)
    at Object.<anonymous> (/mnt/e/User/Music/Spotify/spotify-dl/node_modules/ora/index.js:11:28)
    at Module._compile (node:internal/modules/cjs/loader:1109:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1138:10)
    at Module.load (node:internal/modules/cjs/loader:989:32)
    at Function.Module._load (node:internal/modules/cjs/loader:829:14)
    at Module.require (node:internal/modules/cjs/loader:1013:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/mnt/e/User/Music/Spotify/spotify-dl/node_modules/ora/index.js',
    '/mnt/e/USer/Music/Spotify/spotify-dl/cli.js'
  ]
}
```

I'm not well versed in Git, so this could be completely my own fault. The steps I took were:
Git clone repo and cd to it
gh pr checkout 84
sudo npm install
sudo npm link

Am I missing some steps? I am using WSL.

**Comment by phyzical** [comment]
hmm i dont think so, although i use docker. i just tried my branch and it seemed to work.

I will try on a WSL setup later tonight, 

few things though. 
* why use `sudo npm` if that isnt a WSL thing i would try to avoid adding root privs to npm things
* when you say `checkout 84` just wanted to confirm you actually checkout the branch and not a branch called "84" :) i.e `phyzical:feature/adding-support-for-artist` from my fork of this repo `https://github.com/phyzical/spotify-dl`

Thanks for giving it a try though!

image

**Comment by SwapnilSoni1999** [comment]
I'm still seeing this newline logging and not accurate download% count it always prints in new line and in round figures and the final 100% value as it finishes the download.
image

File: util/get-link.js [hunk 3]
```diff
@@ -5,33 +5,32 @@ const youtubeSearch = require('yt-search');
...
+const getLink = async songName => {
+  const tryLink = async () => {
     const result = await search(songName);
-    
-    const [topResult] = result.videos;
-
-    const youtubeLink = buildUrl(topResult)
-
+    const topResult = result.videos
+      .find(video => video.seconds < (MAX_MINUTES * 60));
```
Comment (SwapnilSoni1999) [inline util/get-link.js:25]: As i mentioned in the comment. Is this the reason why it skips the entire song? can we loop to first 3 songs ? and with a 'lyrics' keyword search as well? Just giving some idea. Please handle the implementation in your preferred way :)

**Review by SwapnilSoni1999 — CHANGES_REQUESTED** [review]

File: util/get-link.js (hunk 3 above)
Comment (phyzical) [reply util/get-link.js:25]: hmm, ill add a comment of "no matches found for" that way we can confirm that's whats occuring,

but the way i was trying to design it was it should get the first song that is under an hour long

**Comment by phyzical** [comment]
hmm @SwapnilSoni1999 the not accurate download is because if we don't then it logs ALOT,

if you remove this line in downloader `if (toBeDownloadedMb % 1 == 0 || toBeDownloadedMb == downloadedMb) {` you will see what i mean. 
Oh and i had to make it log new line due to non interactive consoles i.e docker, crons ect.

Maybe what we can do is detect if its a interactive usage and ill log like i am for non interactive and fall back to what it was doing for interactive?

**Comment by SwapnilSoni1999** [comment]
> hmm @SwapnilSoni1999 the not accurate download is because if we don't then it logs ALOT,
> 
> if you remove this line in downloader `if (toBeDownloadedMb % 1 == 0 || toBeDownloadedMb == downloadedMb) {` you will see what i mean.
> Oh and i had to make it log new line due to non interactive consoles i.e docker, crons ect.
> 
> Maybe what we can do is detect if its a interactive usage and ill log like i am for non interactive and fall back to what it was doing for interactive?

But still one line realtime updated logging is what I am looking for :) please can you fix that and revert

**Comment by phyzical** [comment]
image

there is 100% no skipping of the first few songs.

i was able to get the context of tty so it should now update on  single line like you wanted but also log multi line for docker usages.

i made a few adjustments to the cachefile and did some groundwork for some future work aswell as a few more refactors

**Review by SwapnilSoni1999 — APPROVED** [review]

**Comment by SwapnilSoni1999** [comment]
Awesome work man! Applause :clap: :champagne: 
and lots of respect <3

**Comment by phyzical** [comment]
happy to 👍 will start on the other new features this weekend and i think itll then be feature complete (in my eyes)

**Comment by SwapnilSoni1999** [comment]
> happy to  will start on the other new features this weekend and i think itll then be feature complete (in my eyes)

Awesome. :) take your time man. You're great

**Pull request merged**